### ISSUE TITLE
team_scratch memory grows a lot more since "resize_team_scratch_space" was refactored.

### ISSUE THREAD

**Issue by alanw0** [kokkos/kokkos#4665]
The kokkos commit 31a136c604 refactored the function 'resize_team_scratch_space'. This caused some previously-passing exawind tests to start crashing/failing with a cuda allocation failure.

Here are some print-outs of a (successful) run prior to that commit:
(The 'before' and 'after' is just printouts before and after the kokkos::parallel_for in which the team-scratch memory is being used.)

```
 face-elem before:  bytes_per_thread:13856 free:29326049280
 face-elem after:  bytes_per_thread:13856 free:27388280832
 face-elem before:  bytes_per_thread:13856 free:27388280832
 face-elem after:  bytes_per_thread:13856 free:27388280832
 edge before:  bytes_per_thread:384 free:27384086528
 edge after:  bytes_per_thread:384 free:27384086528
...
...
 edge before:  bytes_per_thread:64 free:26910130176
 edge after:  bytes_per_thread:64 free:26910130176
 edge before:  bytes_per_thread:64 free:26813661184
 edge after:  bytes_per_thread:64 free:26813661184
 after Kokkos::finalize_all():  cuda free:33291763712 cuda total:34089730048
```
'free:' is the amount of free memory on the device, and it hovers around 26 or 27GB throughout the run.
Next are the same printouts from a run after commit 31a136c604:
```
 face-elem before:  bytes_per_thread:13856 free:28726263808
 face-elem after:  bytes_per_thread:13856 free:26694123520
 face-elem before:  bytes_per_thread:13856 free:26694123520
 face-elem after:  bytes_per_thread:13856 free:24599068672
 edge before:  bytes_per_thread:384 free:24594874368
 edge after:  bytes_per_thread:384 free:24594874368
...
...
 edge before:  bytes_per_thread:64 free:10577510400
 edge after:  bytes_per_thread:64 free:10577510400
 edge before:  bytes_per_thread:64 free:10615259136
 edge after:  bytes_per_thread:64 free:10615259136
 after Kokkos::finalize_all():  cuda free:33300152320 cuda total:34089730048
```
As you can see, now the amount of free memory on the device goes down to 10 or 11GB before the run is over. This was run on a 32GB card so it succeeded, but our nightly tests run on a 16GB card and they started crashing because of this.

It isn't really a memory leak, since kokkos::finalize does clean it up at the end of the run, but the excessive growth of the team scratch space has the same effect as a memory leak when it causes us to crash on a card that we used to run successfully on.

We noticed the tests start crashing when the kokkos update was put into trilinos at the end of October. It took us a while to track it down to the kokkos commit above.

Tagging other folks for their information: @ldh4 @jrood-nrel @tasmith4

**Comment by Rombur** [comment]
The commit should only change the behavior if you launch kernels from different threads. If you launch all your kernels from the same thread, there shouldn't be any difference. The scratch memory is associated to an instance not to a `parallel_for` and that's why the memory is released when `finalize` is called but that's not a new behavior. What the commit does is creating a pool of scratch memory that grows if multiple `parallel_for`/`parallel_reduce` require scratch memory simultaneously. If they release the memory before another one tries to acquire it, the memory pool won't grow.

**Comment by alanw0** [comment]
We are not using threads. Kokkos::OpenMP is not enabled in this build.
As the print statements demonstrate, the behavior changed dramatically due to this commit. Same build/configuration, only changed by applying this kokkos commit.

**Comment by Rombur** [comment]
My guess is that there is some kind of "drift" in the memory pool. Can  you change the code [here](https://github.com/kokkos/kokkos/commit/31a136c604708f291a74e43db7e457417fac2aec#diff-299ad949bd0a1700160902ccf49827e7f1ca3bd0444c35ab10438dc24cca925fR661-R662) from
```C++
 while (m_team_scratch_pool[current_team_scratch].compare_exchange_weak(
      zero, one, std::memory_order_release, std::memory_order_relaxed)) {
```
to
```C++
 while (!m_team_scratch_pool[current_team_scratch].compare_exchange_weak(
      zero, one, std::memory_order_release, std::memory_order_relaxed)) {
```

**Comment by alanw0** [comment]
@Rombur That appears to fix it, thank you!
I guess we can apply this as a patch in our installation, until you get it pushed in and get an update into trilinos?

**Comment by Rombur** [comment]
Yes, for now you should apply the patch yourself. We will have a new release of Kokkos in the next few weeks so Trilinos should be update pretty soon but it might still take a month or so.

**Comment by crtrott** [comment]
Crosslink #4673

**Comment by PhilMiller** [comment]
Fix released in 3.6
